src/System.Management.Automation/engine/parser/Compiler.cs

```diff
@@ -3217,37 +3217,120 @@ private void CompileTrappableExpression(List<Expression> exprList, StatementAst
             var expr = Compile(stmt);
             exprList.Add(expr);
 
-            var pipeAst = stmt as PipelineAst;
-            if (pipeAst != null)
+            if (MustSetSuccessAfterEvaluating(stmt))
             {
-                if (pipeAst.PipelineElements.Count == 1 && pipeAst.PipelineElements[0] is CommandExpressionAst)
-                {
-                    // A single expression - must set $? after the expression.
-                    exprList.Add(s_setDollarQuestionToTrue);
-                }
+                exprList.Add(s_setDollarQuestionToTrue);
             }
-            else
+        }
+
+        /// <summary>
+        /// Determines whether a statement must have an explicit setting
+        /// for $? = $true after it by the compiler.
+        /// </summary>
+        /// <param name="statementAst">The statement to examine.</param>
+        /// <returns>True is the compiler should add the success setting, false otherwise.</returns>
+        private bool MustSetSuccessAfterEvaluating(StatementAst statementAst)
```

Comment:
Can we reword the names? "Must" confuses me.

Reply:
Ah! I did play with a few names. Not sure what's best. Please let me know if you have suggestions. Some possibilities:

- `ExpressesPureValue` (didn't like this since it doesn't apply perfectly; this method group is specific to the `$?` optimisation)
- `RequiresDollarHookSet` (uses the nonstandard "dollar hook" name and doesn't describe as fully)
- `IsEvaluatedWithoutPipeline`
- `AlwaysSucceeds`
- `RequiresSuccessSetAfterEvaluating`

Reply:
Docs describe $? as "Last command status", so maybe this could be called `ShouldSetLastCommandStatus`, since that is what it is really evaluating.

Reply:
Maybe `ShouldSetExecutionStatusToSuccess`.

Reply:
Changed to @daxian-dbw's suggestion